src/built_ins.md

```diff
@@ -18,222 +18,757 @@ The built-in types for `alan` are:
-  This function looks for the first value in the array equal to the provided value and returns that index, or `-1`.
-* `length(Array<any>): int64`
-  Returns the number of elements in the Array.
-
-A special function just for `Array`s and `string`s:
-
-* `join(Array<string>, string): string`
-  This function takes the elements of the array, which are all strings, and concatenates them together with the second argument, a separator string, in between them, and then returns the resulting string.
-
-The functional programming-inspired `Array` functions:
-
-* `each(Array<any>, fn (any): void): void`
-  This function calls a side-effect function for each element of the array.
-* `map(Array<any>, fn (any): any): Array<any>`
-  This function allows an array to be transformed into another array with a different set of data after being given a transform function. Coming Soon (tm) more advanced interfaces will provide more solid guarantees that the return type of the transform function matches the return type of the map call itself (needed for the compiler, the interpreter works without it because values are stored dynamically).
-* `reduce(Array<any>, fn (any, any): any): any`
-  This function takes an array and an function that takes two arguments (so equivalent to an infix operator) and merges all of the values together until a single output is provided.
-* `filter(Array<any>, fn (any): bool): Array<any>`
-  This function takes an array and a filter function that specifies which elements should be kept (`true`) and which should be skipped (`false`).
-* `find(Array<any>, fn(any): bool): any`
-  This function takes an array and a fitler function and returns the first value that passes the test of the filter function.
-* `every(Array<any>, fn(any): bool): bool`
-  This function takes an array and a filter function and indicates if every field in the array passed the test (`true`) or not (`false`)
-* `some(Array<any>, fn(any): bool): bool`
-  This function takes an array and a filter function and indicates if any field in the array passed the test or not.
+```rust
+// Takes two values of the same type and returns an array of those two values
+pair(any, any): Array<any>
+```
+
+```rust
+// Takes a boolean and an array of two values and returns the first value on `true` and the second
+// on `false`
+cond(bool, Array<any>): any
+```
+
+```rust
+// Takes a boolean and a function and conditionally executes that function if the bool is `true`
+cond(bool, function): void
+```
+
+##### Assign Functions
+
+```rust
+// Explicitly duplicates the provided value and returns it
+assign(any): any
+assign(Array<any>): Array<any>
+assign(void): void
+assign(int8): int8
+assign(int16): int16
+assign(int32): int32
+assign(int64): int64
+assign(float32): float32
+assign(float64): float64
+assign(bool): bool
+assign(string): string
+```
+
+##### Error, Maybe, Result, and Either Functions
+
+```rust
+// Returns a non-error error
+noerr(): Error
```

Comment:
nit not clear when this would be necessary. also perhaps just an empty error instead of a non-error error? That seems confusing, is it just an empty object?

Reply:
It can't be an empty object since the Error object requires a `msg` string, so it's literally `'No Error'` :)

I can change this to "empty error" if you want, but it's not actually empty.

Reply:
got it. I am ok with either